src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -37,35 +35,45 @@ internal sealed partial class Http2Connection : HttpConnectionBase, IDisposable
         private readonly Dictionary<int, Http2Stream> _httpStreams;
 
         private readonly CreditManager _connectionWindow;
-        private readonly CreditManager _concurrentStreams;
 
         private int _nextStream;
         private bool _expectingSettingsAck;
         private int _initialWindowSize;
-        private int _maxConcurrentStreams;
         private int _pendingWindowUpdate;
         private long _idleSinceTickCount;
 
+        private uint _maxConcurrentStreams;
+        private uint _streamsInUse;
+        private TaskCompletionSource<bool>? _availableStreamsWaiter;
+
         private readonly Channel<WriteQueueEntry> _writeChannel;
         private bool _lastPendingWriterShouldFlush;
 
-        // This means that the pool has disposed us, but there may still be
-        // requests in flight that will continue to be processed.
-        private bool _disposed;
+        // This will be set when we receive a GOAWAY.
+        // Requests currently in flight will continue to be processed, but no new requests can be initiated.
+        private bool _goingAway;
 
-        // This will be set when:
-        // (1) We receive GOAWAY -- will be set to the value sent in the GOAWAY frame
-        // (2) A connection IO error occurs -- will be set to int.MaxValue
-        //     (meaning we must assume all streams have been processed by the server)
-        private int _lastStreamId = -1;
+        // If this is set, the connection is aborting due to an IO failure (IOException) or a protocol violation (Http2ProtocolException).
+        // Requests in flight have been (or are being) failed.
+        private Exception? _abortException;
+
+        // If this is set, the connections is shutting down and cannot accept new requests.
+        // When this is true, one or more of the following is also true:
+        // (1) _goingAway
+        // (2) _abortException is not null
+        // (3) _nextStream == MaxStreamId
+        private bool _shutdown;
+        private TaskCompletionSource? _shutdownWaiter;
+
+        // This means that the pool has disposed us and will not submit further requests.
+        // Requests currently in flight will continue to be processed.
+        // When all requests have completed, the connection will be torn down.
+        private bool _disposed;
```

Comment:
The one that's weirdest, I think, is _goingAway, because it's just one of the reasons we could be shutdown. I'll look at either getting rid of this or folding it into an enum or something.

Aside from that, it's fairly straightforward. We need to track whether we've been disposed locally; that's _disposed. We need to track whether an exogenous shutdown has been initiated; that's _shutdown. Those two are mutually exclusive so combining them into an enum doesn't make much sense, I think.

The only other state, then, is _abortException, which distinguishes between graceful exogenous shutdown vs abortive exogenous shutdown (i.e. IO failure or protocol violation). 

I'll try to make this all clearer...